### DISCUSSION TITLE
How to use a Data Bridge to publish multiple topics?

### DISCUSSION THREAD

**Discussion by matthewGEA** [emqx/emqx#10241] [General]
What is the correct way to use the Data Bridge function in version 5 of emqx to publish multiple topics?  Is there a way for a single bridge to do this?

In previous versions 5.x, a single connection could be defined and then multiple topics could be mapped to this single bridge.  It looks like that this capability has been removed recent versions, defining only a single "egress" or "ingress" object.

**Comment by matthewGEA** [comment]
It looks like this change happened with version 5.0.12

> This version included a refactoring of MQTT bridge config. The older version config file created from v5.0.11 or earlier will be converted to according to the new schema. Please note though, the request body of /bridges API to configure MQTT brdige is changed in a incompatible way.

Prior to 5.0.12, it was possible to use this syntax:
bridges.mqtt.bridge1.connector = connectorID
bridges.mqtt.bridge1.direction = egress
bridges.mqtt.bridge1.local_topic = topic1/#

bridges.mqtt.bridge2.connector = connectorID
bridges.mqtt.bridge2.direction = egress
bridges.mqtt.bridge2.local_topic = topic2/#


Starting with 5.0.12, the syntax is the following:
bridges.mqtt.bridge1.server = serverIPaddress
bridges.mqtt.bridge1.egress.local.topic = topic1/#

bridges.mqtt.bridge2.server = serverIPaddress
bridges.mqtt.bridge2.egress.local.topic = topic2/#

Which this configuration appears to double the number of clients on the remote broker.

**Comment by lafirest** [comment]
@matthewGEA 
<img width="1199" alt="image" src="https://user-images.githubusercontent.com/22971499/227833498-a02b9b2a-3055-45a3-9e56-d332285d283c.png">

<img width="1575" alt="image" src="https://user-images.githubusercontent.com/22971499/227833219-3cbfcb3d-4bc4-47dd-a897-d0d051fe4531.png">

As these pictures showed, you can use different  `Rules`  as upstream to forward various topics to the same bridge.

**Reply by matthewGEA** [reply]
Can you explain the syntax of the configuration setting through environmental variables to link a rule to a Data Bridge?

In the documentation, there are two built-in functions (console and republish), and only a mention of Data Bridge but no details on how to specify this.

I'm looking for a simple example to define: `EMQX_RULE_ENGINE__RULES__$ID` per https://www.emqx.io/docs/en/v5.0/admin/cfg.html#rule-engine-rules

**Reply by lafirest** [reply]
@matthewGEA 
You can find out how to configure each feature via the schema documentation, follow the `POST /rules` example at https://www.emqx.io/docs/en/v5.0/admin/api-docs.html#tag/Rules/paths/~1rules/post, we can see the bridge is linked with rules by its id

**Reply by matthewGEA** [reply]
Thank you for that reference to the documentation.

Does the Bridge need to have an egress rule specified for the egress rules to work?  For example:
`EMQX_bridges__mqtt__bridge__egress__local__topic="t/#"`

I am finding that if I do not have an egress rule set on the bridge, then I get these errors on startup, and then the packets never get published out the bridge:

```
[error] msg: forwarding_unavailable, mfa: emqx_connector_mqtt_worker:export_msg/2, line: 384, message: #{
...
}, reason: egress is not configured
```

**Reply by lafirest** [reply]
There are some modifies among different versions, you can check the detailed schema by starting a local EMQX and opening this URL `http://127.0.0.1:18083/api-docs/index.html`

**Comment by kjellwinblad** [comment]
@matthewGEA did @lafirest comment, answer your question? In this case, can we mark this as answered?

**Comment by matthewGEA** [comment]
Below is an example configuration for a Docker Compose file.

One of the key configurations not explicit in the documentation is how to reference the bridge in the `EMQX_RULE_ENGINE__RULES__rule-one__actions` configuration item.  It needs to be a list (enclosed by square brackets, and needs to have "mqtt:" prior to the name of the bridge.

```
      ### Setup the bridge MQTT connection information
      - EMQX_bridges__mqtt__bridge__enable=true
      - EMQX_bridges__mqtt__bridge__server="${UPSTREAMSERVERADDR}:1883"
      - EMQX_bridges__mqtt__bridge__clientid_prefix="${GROUPID}-spb"
      - EMQX_bridges__mqtt__bridge__proto_ver=v5
      - EMQX_bridges__mqtt__bridge__bridge_mode=false
      - EMQX_bridges__mqtt__bridge__username=${USERNAME}
      - EMQX_bridges__mqtt__bridge__password=${PASSWORD}
      - EMQX_bridges__mqtt__bridge__keepalive=300s
      - EMQX_bridges__mqtt__bridge__retry_interval=15s
      - EMQX_bridges__mqtt__spb__max_inflight=32

      ### Use the Rules Engine to forward matching packets to the MQTT Bridge
      - EMQX_RULE_ENGINE__RULES__rule-one__name="NBIRTH"
      - EMQX_RULE_ENGINE__RULES__rule-one__sql="SELECT * FROM \"t/#\""
      - EMQX_RULE_ENGINE__RULES__rule-one__actions=["mqtt:bridge"]
      - EMQX_RULE_ENGINE__RULES__rule-one__enable="true"
      - EMQX_RULE_ENGINE__RULES__rule-one__description="Forward t topic data"
```
